qiskit/backends/_sympy_unitarysimulator.py

```diff
@@ -0,0 +1,280 @@
+
+"""Contains a (slow) Python simulator that returns the unitary of the circuit.
+
+It produces the unitary of a quantum circuit in the symbolic form.
+In particular, it simulates the quantum computation with the sympy APIs,
+which preserve the symbolic form of numbers, e.g., sqrt(2), e^{i*pi/2}.
+
+How to use this simulator:
+see examples/python/use_sympy_simulators.py
+
+Example output:
+[[sqrt(2)/2 sqrt(2)/2 0 0]
+ [0 0 sqrt(2)/2 -sqrt(2)/2]
+ [0 0 sqrt(2)/2 sqrt(2)/2]
+ [sqrt(2)/2 -sqrt(2)/2 0 0]]
+
+Warning: it is slow.
+"""
+import uuid
+import logging
+import numpy as np
+from sympy.matrices import eye, zeros
+from sympy.physics.quantum import TensorProduct
+from sympy import Matrix, pi, E, I, cos, sin, N
+from qiskit._result import Result
+from qiskit.backends._basebackend import BaseBackend
+
+logger = logging.getLogger(__name__)
+
+
+def index1(barg, iarg, karg):
+    """Magic index1 function.
+
+    Takes a bitstring k and inserts bit b as the ith bit,
+    shifting bits >= i over to make room.
+    """
+    retval = karg
+    lowbits = karg & ((1 << iarg) - 1)  # get the low i bits
+
+    retval >>= iarg
+    retval <<= 1
+
+    retval |= barg
+
+    retval <<= iarg
+    retval |= lowbits
+
+    return retval
+
+
+def index2(barg1, iarg1, barg2, iarg2, k):  # assert(i1 != i2)
+    """Magic index1 function.
+
+    Takes a bitstring k and inserts bits b1 as the i1th bit
+    and b2 as the i2th bit
+    """
+    if iarg1 > iarg2:
+        # insert as (i1-1)th bit, will be shifted left 1 by next line
+        retval = index1(barg1, iarg1 - 1, k)
+        retval = index1(barg2, iarg2, retval)
+    else:  # i2>i1
+        # insert as (i2-1)th bit, will be shifted left 1 by next line
+        retval = index1(barg2, iarg2 - 1, k)
+        retval = index1(barg1, iarg1, retval)
+    return retval
+
+
+class SympyUnitarySimulator(BaseBackend):
+    """Python implementation of a unitary simulator."""
+
+    def __init__(self, configuration=None):
+        """Initial the UnitarySimulator object."""
+        super(SympyUnitarySimulator, self).__init__()
+        if configuration is None:
+            self._configuration = {'name': 'local_sympy_unitary_simulator',
+                                   'url': 'https://github.com/QISKit/qiskit-sdk-py',
+                                   'simulator': True,
+                                   'local': True,
+                                   'description': 'A python simulator for unitary matrix',
+                                   'coupling_map': 'all-to-all',
+                                   'basis_gates': 'u1,u2,u3,cx,id'}
+        else:
+            self._configuration = configuration
+        self._unitary_state = None
+        self._number_of_qubits = None
+
+    # the following is the sympy implementations of various gates, including basic gates and ugates.
+    @staticmethod
+    def regulate(theta):
+        """Find the most close symbolic form"""
+        error_margin = 0.01
```

Comment:
The margin seams too low.

Reply:
regulate is trying to recover the symbolic form of pi/N (used as parameters in u gates) because the qiskit will convert pi/N to floating point. I guess the precision loss due to floating point wont exceed 0.01 unless you amplify the loss with a big number, which is unlikely for the u gate parameters.

Reply:
compile converts symbols to float. That's a mistake. I will take a look to it.

Reply:
that is my memory from 2017, not sure whether it still holds. 
Please check it and let me know. thanks.